<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Javascript Errors</title>
</head>
<body>
    <h1>Javascript Errors</h1>
    <button onclick="error()">Error</button>
    <button onclick="syntaxError()">Syntax Error</button>
    <button onclick="typeError()">Type Error</button>
    <hr>
    <p>Please enter a number between 1 and 100.</p>
    <input type="text" id="input">
    <button onclick="validate()">Validate</button>
    <script>
        function error(){
            const message = new Error('Custom Error');
            throw message;
        }
        function syntaxError(){
            throw new SyntaxError('Custom SyntaxError');
        }
        function typeError(){
            throw new TypeError('Custom TypeError');
        }
        try {
            throw new Error('Custom Error');
        } catch (err){
            if (err instanceof Error){
                console.log('instanced');
            }
            console.log(err);
        }

        function validate(value){
            value = value.trim();
            if (value === ''){
                throw 'Boş alan';
            } else if (isNaN(value)){
                throw 'Değer bir sayı değil.';
            }
            value = Number(value);
            if (value < 5){
                throw "Sayı 5'ten küçük olamaz."
            } else if (value > 100){
                throw "Sayı 100'den büyük."
            }

            return true;
        }
       function test (){
        try {
            const value = document.getElementById('input').value;
            validate(value);
        } catch (err){
            console.log(err);
        }
    }

    // Birkaç adet örnek

    // Syntax Error
    // console.log(Hello World); // Burada tırnaklar eksik, SyntaxError oluşur

    // Runtime Error
    let numbers = [1, 2, 3];
    // console.log(numbers[10]); // Burada numbers dizisinin sınırlarını aşıldığı için TypeError oluşur

    // Logic Error
    function calculateSum(a, b) {
   // return a  b; // Burada toplama işlemi yerine çıkarma yapıldığı için beklenmeyen sonuç oluşabilir
    }

    // Try-Catch blokları
    try {
     // Potansiyel hata üretebilecek kodlar
    let x = y + 1; // Burada y tanımlı değilse ReferenceError oluşur
    } catch (error) {
    // Hata durumunda yapılacak işlemler
    //console.error('Error occurred:', error.message);
    }

    // Throwing Errors
    function divide(a, b) {
    if (b === 0) {
    throw new Error('Division by zero is not allowed');
     }
    return a / b;
    }
    // Global Error Handling
    window.onerror = function(message, source, lineno, colno, error) {
     console.error('Global error occurred:', message, 'at', source, 'line', lineno);
    };

    </script>
</body>
</html>